cpp/src/strings/utilities.cu

```diff
@@ -145,20 +145,42 @@ std::unique_ptr<column> make_empty_strings_column( rmm::mr::device_memory_resour
 namespace
 {
 
-std::mutex g_flags_table_mutex;
-static character_flags_table_type* d_character_flags_table = nullptr;
+// guaranteed thread-safe
+struct character_flags_singleton
+{
+    static character_flags_singleton& instance()
+    {
+        static character_flags_singleton _instance;
+        return _instance;
+    }
+    character_flags_singleton( character_flags_singleton const& ) = delete;
+    character_flags_singleton( character_flags_singleton&& ) = delete;
+    character_flags_singleton& operator=(character_flags_singleton const&) = delete;
+    character_flags_singleton& operator=(character_flags_singleton&&) = delete;
 
-} // namespace
+    character_flags_table_type* table() { return d_character_flags_table; }
 
-const character_flags_table_type* get_character_flags_table()
-{
-    std::lock_guard<std::mutex> guard(g_flags_table_mutex);
-    if( !d_character_flags_table )
+private:
+
+    character_flags_table_type* d_character_flags_table{};
+
+    character_flags_singleton()
     {
         RMM_TRY(RMM_ALLOC(&d_character_flags_table,sizeof(g_character_codepoint_flags),0));
```

Comment:
Is this a different singleton than in the other PR? Let's avoid singletons.

Reply:
This is from a different PR #3159  and should be fixed there now.